### ISSUE TITLE
Student_attendance_rf_model.pkl

### ISSUE THREAD

**Issue by chetangangaji** [chetangangaji/Attendance_Prediction#1]
# ============================================================
# STUDENT ATTENDANCE PREDICTION
# RANDOM FOREST CLASSIFIER
# ============================================================

import pandas as pd
import joblib

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# ============================================================
# 1. LOAD DATASET
# ============================================================

df = pd.read_csv("Attendance_Prediction_2000.csv")

print("Dataset loaded successfully!")
print("Dataset shape:", df.shape)
print("\nFirst 5 rows:")
print(df.head())

# ============================================================
# 2. SEPARATE FEATURES AND TARGET
# ============================================================

X = df.drop(columns=["student_id", "attendance", "absence_reason"])
y = df["attendance"]

# ============================================================
# 3. IDENTIFY CATEGORICAL AND NUMERICAL COLUMNS
# ============================================================

categorical_cols = X.select_dtypes(include=["object"]).columns.tolist()
numerical_cols = X.select_dtypes(
    include=["int64", "float64"]
).columns.tolist()

print("\nCategorical columns:", categorical_cols)
print("Numerical columns:", numerical_cols)

# ============================================================
# 4. PREPROCESSING
# ============================================================

preprocessor = ColumnTransformer(
    transformers=[
        ("num", "passthrough", numerical_cols),

        (
            "cat",
            OneHotEncoder(
                drop="first",
                handle_unknown="ignore"
            ),
            categorical_cols
        )
    ]
)

# ============================================================
# 5. RANDOM FOREST MODEL
# ============================================================

model_pipeline = Pipeline(
    steps=[
        ("preprocessor", preprocessor),

        (
            "classifier",
            RandomForestClassifier(
                n_estimators=100,
                max_depth=10,
                random_state=42
            )
        )
    ]
)

# ============================================================
# 6. TRAIN-TEST SPLIT
# ============================================================

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("\nTraining samples:", len(X_train))
print("Testing samples:", len(X_test))

# ============================================================
# 7. TRAIN MODEL
# ============================================================

print("\nTraining the model...")

model_pipeline.fit(X_train, y_train)

print("Model trained successfully!")

# ============================================================
# 8. EVALUATE MODEL
# ============================================================

y_pred = model_pipeline.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)

print("\n==============================")
print("MODEL PERFORMANCE")
print("==============================")

print(f"Accuracy: {accuracy:.4f}")
print(f"Accuracy Percentage: {accuracy * 100:.2f}%")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# ============================================================
# 9. SAVE MODEL
# ============================================================

joblib.dump(
    model_pipeline,
    "student_attendance_rf_model.pkl"
)

print("\nModel saved successfully!")
print("File: student_attendance_rf_model.pkl")

# ============================================================
# 10. TAKE REAL-WORLD USER INPUT
# ============================================================

print("\n")
print("==========================================")
print(" STUDENT ATTENDANCE PREDICTION")
print("==========================================")

age = int(input("Enter Age: "))

gender = input(
    "Enter Gender (male/female): "
).lower()

course = input(
    "Enter Course (bca/bsc/bcom/ba/bba): "
).lower()

year = input(
    "Enter Year (1st year/2nd year/3rd year): "
).lower()

parent_education = input(
    "Enter Parent Education (school/graduate/postgraduate): "
).lower()

internet_access = input(
    "Internet Access (yes/no): "
).lower()

hostel_resident = input(
    "Hostel Resident (yes/no): "
).lower()

class_type = input(
    "Class Type (offline/online): "
).lower()

weather = input(
    "Weather (sunny/cloudy/rainy): "
).lower()

study_hours = float(
    input("Study Hours per day: ")
)

sleep_hours = float(
    input("Sleep Hours per day: ")
)

travel_time_minutes = int(
    input("Travel Time in minutes: ")
)

# ============================================================
# 11. CREATE USER INPUT DATAFRAME
# ============================================================

sample_student = pd.DataFrame(
    [
        {
            "age": age,
            "gender": gender,
            "course": course,
            "year": year,
            "parent_education": parent_education,
            "internet_access": internet_access,
            "hostel_resident": hostel_resident,
            "class_type": class_type,
            "weather": weather,
            "study_hours": study_hours,
            "sleep_hours": sleep_hours,
            "travel_time_minutes": travel_time_minutes
        }
    ]
)

# ============================================================
# 12. PREDICT ATTENDANCE
# ============================================================

prediction = model_pipeline.predict(sample_student)[0]

probability = model_pipeline.predict_proba(
    sample_student
)[0]

# ======================
# ============================================================
# 13. DISPLAY RESULT
# ============================================================

print("\n")
print("==========================================")
print("          PREDICTION RESULT")
print("==========================================")

print(f"Model Accuracy: {accuracy * 100:.2f}%")

if prediction == 1:
    print("Predicted Attendance: PRESENT")
else:
    print("Predicted Attendance: ABSENT")

print(f"Absent Probability: {probability[0] * 100:.2f}%")
print(f"Present Probability: {probability[1] * 100:.2f}%")

print("==========================================")
